packages/test/test-end-to-end-tests/src/test/refreshSerializerLifeCycle.spec.ts

```diff
@@ -0,0 +1,198 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { strict as assert } from "assert";
+
+import { generatePairwiseOptions } from "@fluid-private/test-pairwise-generator";
+import { describeCompat } from "@fluid-private/test-version-utils";
+import { Deferred } from "@fluidframework/core-utils/internal";
+import {
+	ITestFluidObject,
+	timeoutPromise,
+	DataObjectFactoryType,
+	createAndAttachContainer,
+	timeoutAwait,
+	waitForContainerConnection,
+	type ChannelFactoryRegistry,
+	type ITestObjectProvider,
+} from "@fluidframework/test-utils/internal";
+import type { IContainerExperimental } from "@fluidframework/container-loader/internal";
+import {
+	type IContainerRuntimeOptions,
+	DefaultSummaryConfiguration,
+} from "@fluidframework/container-runtime/internal";
+import type {
+	ConfigTypes,
+	IConfigProviderBase,
+	ITelemetryBaseLogger,
+} from "@fluidframework/core-interfaces/internal";
+import { SharedMap, type ISharedMap } from "@fluidframework/map/internal";
+import { MockLogger } from "@fluidframework/telemetry-utils/internal";
+import type { IHostLoader } from "@fluidframework/container-definitions/internal";
+import { wrapObjectAndOverride } from "../mocking.js";
+
+const testConfigs = generatePairwiseOptions({
+	savedOps: [true, false],
+	pendingOps: [true, false],
+	savedOps2: [true, false],
+	pendingOps2: [true, false],
+	summaryWhileOffline: [true, false],
+	waitForRefresh: [true,false],
+});
+
+describeCompat("Validate Attach lifecycle", "NoCompat", (getTestObjectProvider, apis) => {
+	const mapId = "map";
+	const registry: ChannelFactoryRegistry = [[mapId, SharedMap.getFactory()]];
+	const configProvider = (settings: Record<string, ConfigTypes>): IConfigProviderBase => ({
+		getRawConfig: (name: string): ConfigTypes => settings[name],
+	});
+	const runtimeOptions: IContainerRuntimeOptions = {
+		summaryOptions: {
+			summaryConfigOverrides: {
+				...DefaultSummaryConfiguration,
+				...{
+					maxTime: 5000 * 12,
+					maxAckWaitTime: 120000,
+					maxOps: 1,
+					initialSummarizerDelayMs: 20,
+				},
+			},
+		},
+		enableRuntimeIdCompressor: "on",
+	};
+
+	const waitForSummary = async (container) => {
+		await timeoutPromise((resolve, reject) => {
+			let summarized = false;
+			container.on("op", (op: { type: string; }) => {
+				if (op.type === "summarize") {
+					summarized = true;
+				} else if (summarized && op.type === "summaryAck") {
+					resolve();
+				} else if (op.type === "summaryNack") {
+					reject(new Error("summaryNack"));
+				}
+			});
+		});
+	};
+
+	for (const testConfig of testConfigs) {
+		let provider: ITestObjectProvider;
+		let mockLogger: MockLogger;
+		let getLatestSnapshotInfoP: Deferred<void>;
+		let loader: IHostLoader;
+		let container: IContainerExperimental;
+		let url;
+		let map: ISharedMap;
+		let dataStore: ITestFluidObject;
+		let testContainerConfig;
+		beforeEach(async () => {
```

Comment:
before each doesn't need to be inside the loop. I also dislike using before each, and bunch of global mutable state, as it makes the test much harder to understand and maintain. I'd prefer if the reused code was put into free function, and as much state as possible was made const and allocated within the test function itself.